predicators/approaches/grammar_search_invention_approach.py

```diff
@@ -875,3 +887,92 @@ def _get_successors(
         score_function.evaluate(kept_predicates)  # log useful numbers
 
         return set(kept_predicates)
+
+    def _select_predicates_by_clustering(
+            self, candidates: Dict[Predicate, float],
+            initial_predicates: Set[Predicate], dataset: Dataset,
+            atom_dataset: List[GroundAtomTrajectory]) -> Set[Predicate]:
+        """Cluster segments from the atom_dataset into clusters corresponding
+        to operators and use this to select predicates."""
+
+        if CFG.grammar_search_pred_clusterer == "oracle":
+            assert CFG.offline_data_method == "demo+gt_operators"
+            assert dataset.annotations is not None and len(
+                dataset.annotations) == len(dataset.trajectories)
+            assert CFG.segmenter == "option_changes"
+            segmented_trajs = [
+                segment_trajectory(traj) for traj in atom_dataset
+            ]
+            assert len(segmented_trajs) == len(dataset.annotations)
+            # First, get the set of all ground truth operator names.
+            all_gt_op_names = set(ground_nsrt.parent.name
```

Comment:
hopefully operator names are unique, but I'm not sure if we actually assume/enforce that anywhere. to be safe, can we use the operators themselves instead of their names here?